src/logging.cpp

```diff
@@ -29,45 +34,169 @@ bl::sources::severity_logger<int> warning(3);  // Strange events
-    << "    --version                 | print the version of sunshine"sv << std::endl
-    << std::endl
-    << "    flags"sv << std::endl
-    << "        -0 | Read PIN from stdin"sv << std::endl
-    << "        -1 | Do not load previously saved state and do retain any state after shutdown"sv << std::endl
-    << "           | Effectively starting as if for the first time without overwriting any pairings with your devices"sv << std::endl
-    << "        -2 | Force replacement of headers in video stream"sv << std::endl
-    << "        -p | Enable/Disable UPnP"sv << std::endl
-    << std::endl;
-}
+BOOST_LOG_ATTRIBUTE_KEYWORD(severity, "Severity", int)
+
+namespace logging {
+  /**
+   * @brief A destructor that restores the initial state.
+   */
+  deinit_t::~deinit_t() {
+    deinit();
+  }
+
+  /**
+   * @brief Deinitialize the logging system.
+   *
+   * EXAMPLES:
+   * ```cpp
+   * deinit();
+   * ```
+   */
+  void
+  deinit() {
+    log_flush();
+    bl::core::get()->remove_sink(sink);
+    sink.reset();
+  }
+
+  /**
+   * @brief Initialize the logging system.
+   * @param min_log_level The minimum log level to output.
+   * @param log_file The log file to write to.
+   * @returns A deinit_t object that will deinitialize the logging system when it goes out of scope.
+   *
+   * EXAMPLES:
+   * ```cpp
+   * log_init(2, "sunshine.log");
+   * ```
+   */
+  [[nodiscard]] std::unique_ptr<deinit_t>
+  init(int min_log_level, const std::string &log_file) {
+    if (min_log_level >= 1) {
+      av_log_set_level(AV_LOG_QUIET);
+    }
+    else {
+      av_log_set_level(AV_LOG_DEBUG);
+    }
+    av_log_set_callback([](void *ptr, int level, const char *fmt, va_list vl) {
+      static int print_prefix = 1;
+      char buffer[1024];
+
+      av_log_format_line(ptr, level, fmt, vl, buffer, sizeof(buffer), &print_prefix);
+      if (level <= AV_LOG_ERROR) {
+        // We print AV_LOG_FATAL at the error level. FFmpeg prints things as fatal that
+        // are expected in some cases, such as lack of codec support or similar things.
+        BOOST_LOG(error) << buffer;
+      }
+      else if (level <= AV_LOG_WARNING) {
+        BOOST_LOG(warning) << buffer;
+      }
+      else if (level <= AV_LOG_INFO) {
+        BOOST_LOG(info) << buffer;
+      }
+      else if (level <= AV_LOG_VERBOSE) {
+        // AV_LOG_VERBOSE is less verbose than AV_LOG_DEBUG
+        BOOST_LOG(debug) << buffer;
+      }
+      else {
+        BOOST_LOG(verbose) << buffer;
+      }
+    });
+
+    sink = boost::make_shared<text_sink>();
+
+    boost::shared_ptr<std::ostream> stream { &std::cout, NoDelete {} };
+    sink->locked_backend()->add_stream(stream);
+    sink->locked_backend()->add_stream(boost::make_shared<std::ofstream>(log_file));
+    sink->set_filter(severity >= min_log_level);
+
+    sink->set_formatter([message = "Message"s, severity = "Severity"s](const bl::record_view &view, bl::formatting_ostream &os) {
```

Comment:
Consider moving `message` and `severity` as `constexpr` near the `DATE_BUFFER_SIZE` as they serve absolutely no purpose in the capture list other than to confuse you...